### ISSUE TITLE
Niches/Fitness Sharing for PBT; PBT+BO.

### ISSUE THREAD

**Issue by jon-chuang** [ray-project/ray#5630]
As described in: https://deepmind.com/blog/article/how-evolutionary-selection-can-train-more-capable-self-driving-cars, we could

1. **Modify PBT to include niches**. This amounts to adding a config for number of niches (we can also think about variable size niches, but this is too complicated). Then, the code needs to be refactored so that the current PBT is just a special case with n_niches = 1, and quantiles are computed within niche. It seems reasonable that the scheduling still be handled locally, rather than spawning workers.
There is an ambiguity in the blogpost about whether niches themselves can die/get replaced. This seems reasonable, where this process acts on a much slower time scale. An interesting direction to explore is a hierarchy of niches working on different timescales. 
Of course, this is only meaningful if you have cluster size hardware. I imagine about ~100 workers, ~10 niches could be decent. 
For a hierarchy of depth 3, ~1000 workers, ~10 top-level niches, ~10 bottom-level niches, leaving 10 workers per bottom-level niche, seems fine. Of course I have no intuition on this and numbers for genetic algs.

2. **Fitness-sharing**: This is a heuristic that can take many forms in terms of implementation. 
An interesting direction I can see is with Bayesian Opt over *some* set of prior distributions (uniform, log-uniform are most common), where we compute the posterior distribution given the current population, perhaps using Gaussian processes (though not sure how to deal w/ grids) or other standard HPO ways of approximating this dist, perhaps bayesian linear regression w/ learnt basis expansion for higher dimensional hyperparam spaces. Then we treat trainers with out of posterior dist params more leniently. However, this seems weird and unprincipled.
One can also do Bayesian Opt w/ acquisition function which seems more principled.

3. My own idea: **Bayesian PBT?**
Aside: actually, why doesn't anyone do Bayesian opt with PBT? It seems like a very natural thing to do. Same way SH early stopping benefits from BO through BOHB. However, it seems that PBT is a heuristic for bayesian-informed early stopping; we exploit and mutate around regions that show promise. However, it lacks the unexplored-regions awareness part. (This is replaced by random sampling, for instance, 0.25 of the time)
However, in the current tune implementation of PBT, we sample from the same prior when mutating. It seems some improvement can be made here, as PBT will end up reexploring param regions it already knows a lot about but shows little promise. 
The main advantage, in my opinion of BO over random search (or random mutation) is that it provides a sort of "memory" so we are smarter in terms of what we've already seen.

I suppose an argument against this is that we want to learn hyperparam *schedules* rather than optimal hyperparams per se. So it might be worth reexploring regions of the space that did not work in the past. This is also an argument for why we might want to encourage diversity, we need to explore *at every stage of training*. These are actually good arguments for using the Bayesian acquisition function to do exploration, with posterior taken *wrt current set of hyperparams*, rather than over historical params. This is something I am keen on seeing being done.

I would also like to see a more principled analysis of PBT in terms of Bayesian theory and in terms of exploitation/exploration tradeoffs over a training schedule. Is this problem not simply BO with an additional "time/progress" dimension? However, we are limited in that we can only explore forward and can't go back along this dimension. Furthermore, this is just a naive thought. Really, the other dimensions are the population weights, which can be million dimensional, or at least a high-dimensional submanifold of "fit" weights. This fact could suggest that BO might be unhelpful, since one set of hyperparams might work for one network's weights but not another. 

If someone has some relevant literature to point to, this would be immensely helpful. 

**Call to arms**: These doubts, however, can only be resolved empirically; my intuition is that you stand a good chance to see gains if you perform BO at the *niche* level with a posterior computed over the *current population*. This way, you explore efficiently within niche, where niche-level weights are not too different. 
However, I could be entirely wrong; BOHB for instance performs excellently despite retaining a memory over the entire training trajectory; this suggests that optimal hyperparams may not vary too much over the training curve. However, the HP schedule provided in deepmind's Quake paper (fg. 2 in https://arxiv.org/abs/1807.01281) suggests otherwise. 

My guess is that one has to choose a hyper-hyper-parameters suitable to the task at hand; w/ RL with much less stability; one might benefit from local rather than global searches; this suggests that BO-PBT might provide stronger gains for more stable training regimes without weight mutation like GAN training (actually, is weight mutation possible for GANs? Ensembling/Polyak averaging. Seems very reasonable for already data-parallel pipelines.).

The main strength I see in BO is actually the long-term exploration while exploiting, which is not utilised by SH algs like HyperBand. 

Again there is ambiguity in the case of niches; one can compute, rather than the global equivalents, within niche-uniqueness scores/within niche-posteriors.

**Comment by richardliaw** [comment]
@jon-chuang this is very exciting - can you reach out to me via email? you can find my email on GitHub.
